nexus/src/app/sagas/vpc_create.rs

```diff
@@ -315,6 +374,21 @@ async fn svc_update_firewall(
     Ok(rules)
 }
 
+async fn svc_update_firewall_undo(
+    sagactx: NexusActionContext,
+) -> Result<(), anyhow::Error> {
+    let osagactx = sagactx.user_data();
+    let params = sagactx.saga_params::<Params>()?;
+    let opctx = OpContext::for_saga_action(&sagactx, &params.serialized_authn);
+    let (authz_vpc, _) =
+        sagactx.lookup::<(authz::Vpc, db::model::Vpc)>("vpc")?;
+    osagactx
+        .datastore()
+        .vpc_update_firewall_rules(&opctx, &authz_vpc, vec![])
```

Comment:
With the caveat that I don't really understand Sagas yet, this empty vector seems odd to me. Wouldn't we want an undo to restore the old set of rules, or at least the default rules? I don't think we default to an empty set anywhere else.

Reply:
I believe the implementation of `vpc_update_firewall_rules` does the following:

- Deletes all firewall rules attached to the supplied `vpc_id`
- Inserts new rules, which reference `vpc_id`

This is why we're passing the empty vec - it clears out the old rules, and doesn't insert anything new.

Since this is an "undo" step along the pathway of VPC creation, I don't *think* it matters that we chuck the firewall rules -- we're making this call to delete all resources that might reference the VPC before we remove it.

TL;DR: I cleared it out to clean our SQL tables before removing the associated VPC. Does that seem okay?

Reply:
Thank you for that very helpful explanation; I see now. Yes, that seems fine.